s3-fallback-store/src/main/scala/akka/persistence/s3fallback/S3Fallback.scala

```diff
@@ -0,0 +1,529 @@
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model.GetObjectResponse
+import software.amazon.awssdk.services.s3.model.GetObjectRequest
+import software.amazon.awssdk.services.s3.model.HeadObjectRequest
+import software.amazon.awssdk.services.s3.model.NoSuchBucketException
+import software.amazon.awssdk.services.s3.model.NoSuchKeyException
+import software.amazon.awssdk.services.s3.model.PutObjectRequest
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.jdk.CollectionConverters.MapHasAsJava
+import scala.jdk.CollectionConverters.MapHasAsScala
+import scala.jdk.FutureConverters._
+import scala.util.Failure
+import scala.util.Success
+
+import java.time.Instant
+import java.time.Clock
+import akka.actor.typed.Props
+
+final class S3Fallback(system: ActorSystem[_], config: Config, configLocation: String)
+    extends EventFallbackStore[String]
+    with SnapshotFallbackStore[String] {
+  import EventFallbackStore._
+  import SnapshotFallbackStore._
+  import S3Fallback._
+  import system.executionContext
+
+  override def toBreadcrumb(maybeBreadcrumb: AnyRef): Option[String] =
+    maybeBreadcrumb match {
+      case s: String => Some(s)
+      case _         => None
+    }
+
+  override def breadcrumbClass: Class[String] = classOf[String]
+
+  override def close(): Unit = {
+    log.info("Closing S3 fallback")
+    warmer.foreach(_ ! Closing)
+    client.close()
+  }
+
+  override def loadEvent(
+      breadcrumb: String,
+      persistenceId: String,
+      seqNr: Long,
+      includePayload: Boolean): Future[Option[EventFromFallback]] =
+    eventFallbackInitialFuture.flatMap { _ => _loadEvent(breadcrumb, persistenceId, seqNr, includePayload) }
+
+  private def _loadEvent(
+      bucket: String, // the breadcrumb
+      persistenceId: String,
+      seqNr: Long,
+      includePayload: Boolean): Future[Option[EventFromFallback]] = {
+    if (log.isDebugEnabled) {
+      log.debug(s"Retrieving event from bucket [$bucket] for persistenceId [$persistenceId] at seqNr [$seqNr]")
+    }
+
+    val eventFolder = seqNr / eventsPerFolder
+    val key = s"${keyForPid(persistenceId)}/$eventFolder/$seqNr"
+
+    if (includePayload) {
+      val req = GetObjectRequest.builder.bucket(bucket).key(key).build()
+
+      client
+        .getObject(req, AsyncResponseTransformer.toBytes[GetObjectResponse])
+        .asScala
+        .flatMap { respBytes =>
+          val response = respBytes.response
+          // AWS SDK constructs the byte array and promises not to modify
+          val bytes = respBytes.asByteArrayUnsafe
+
+          val metadata = response.metadata.asScala.toMap
+
+          event(Some(bytes), metadata)
+        }
+        .andThen {
+          case Success(_) =>
+            if (log.isDebugEnabled) {
+              log.debug(s"Retrieved event from bucket [$bucket] for persistenceId [$persistenceId] at seqNr [$seqNr]")
```

Comment:
We normally use format templates for debug logs:

```scala
log.debug(
  "Retrieved event from bucket [{}] for persistenceId [{}] at seqNr [{}]",
  bucket,
  persistenceId,
  seqNr)
```

Reply:
Apply to all the `log.debug` with `s"..."`.

Reply:
Thinking there was no need to impose vararg overhead when we're sure that debug is enabled.

Reply:
Ok, sounds good.